core: setup core to support classic confinement.

sergiusens opened the pull request:
When doing ci, most systems won't have /snap/core/current as they
are not snappy capable. The work here allows for one to export
SNAPCRAFT_SETUP_CORE to get the core snap and unsquash it
onto the expected locations.

LP: #1662351

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

sergiusens commented:
This branch still depends on whatever comes out of #940

elopio commented on integration_tests/test_stage.py:
```diff
@@ -64,3 +66,26 @@ def test_staging_libc_links(self):
+
+        # Now we set the required environment variable
+        self.useFixture(fixtures.EnvironmentVariable(
+                'SNAPCRAFT_SETUP_CORE', '1'))
```
so, how do we set this variable in launchpad?

kyrofa replied:
It sounds like cjwatson will be mounting the core snap into place, so perhaps we don't need to worry about setting this in LP.

sergiusens replied:
I already discussed this with him. If they need it, they will set it as they have done with all other env var enabled features (as it is their preference)

elopio commented on snapcraft/__init__.py:
```diff
@@ -282,12 +282,14 @@
 from snapcraft._baseplugin import BasePlugin        # noqa
 from snapcraft._options import ProjectOptions       # noqa
 from snapcraft._help import topic_help              # noqa
+# TODO migrate to snapcraft.store
```
This is the type of things that we read a month from now and we no longer remember why they made sense before and not now.
Can you please explain why, and add the date and your name?

sergiusens replied:
I'll add a bug but name and date are the reason we use a VCS ;-)

elopio commented on snapcraft/internal/cache/_snap.py:
```diff
@@ -76,8 +76,24 @@ def prune(self, *, keep_revision):
+    def get(self, *, snap_name, revision=None):
+        cached_snaps = []
+        for f in os.listdir(self.snap_cache_dir):
+            if f.startswith(snap_name):
```
isn't this risky with snaps whose name is a subset of other snaps? Like hello and hello-world?

sergiusens replied:
this isn't a problem once I merge the new work on delta branches.

elopio commented on snapcraft/internal/lifecycle.py:
```diff
@@ -112,6 +118,39 @@ def execute(step, project_options, part_names=None):
+    if os.path.exists(core_path):
+        logger.debug('{!r} already exists, skipping core setup'.format(
+            core_path))
+        return
```
This if condition deserves a unit test, I think.

elopio commented on snapcraft/internal/lifecycle.py:
```diff
@@ -112,6 +118,39 @@ def execute(step, project_options, part_names=None):
+    # This should now be a hit.
+    core_snap = snap_cache.get(snap_name='core', revision=core_revision)
+    if not core_snap:
+        raise RuntimeError('Error while retrieving the core snap')
```
missing unit test.

elopio commented on snapcraft/internal/lifecycle.py:
```diff
@@ -112,6 +118,39 @@ def execute(step, project_options, part_names=None):
+            # revision here.
+            snap_cache.cache(snap_name, revision=core_revision)
+    # This should now be a hit.
+    core_snap = snap_cache.get(snap_name='core', revision=core_revision)
```
maybe, you could simplify this with a try catch. Make get raise an exception instead of return None.

```
try:
    core_snap = cache.get...
except NotFoundInCache:
    core_snap = download  # Raise if fails to download.
    save_to_cache(core_snap)
```

I don't see why you need to call core_snap twice, and check if not core_snap twice.

elopio commented on snapcraft/internal/lifecycle.py:
```diff
@@ -112,6 +118,39 @@ def execute(step, project_options, part_names=None):
+            snap_name = os.path.join(d, 'core_stable_{}.snap'.format(deb_arch))
+            snapcraft.download('core', 'stable', snap_name, deb_arch)
+            # FIXME there is a chance of incorrecly caching the wrong
+            # revision here.
```
you are not really explaining here why. And if it's a FIXME, it should have a bug reported to fix it later.

kyrofa commented on snapcraft/internal/lxd.py:
```diff
@@ -50,7 +51,8 @@ def _pull_file(self, src, dst):
     def _container_run(self, cmd):
-        check_call(['lxc', 'exec', self._container_name, '--'] + cmd)
+        check_call(['lxc', 'exec', self._container_name, '--'] + cmd,
+                   env=self._env)
```
This sets the environment on the _host_ for this command, doesn't it? Does that get propagated into the `lxc exec`, or should this be set within the command itself?

kyrofa commented on snapcraft/internal/lifecycle.py:
```diff
@@ -112,6 +118,39 @@ def execute(step, project_options, part_names=None):
+    # Now unpack
+    logger.info('Setting up {!r} in {!r}'.format(core_snap, core_path))
+    check_call(['sudo', 'mkdir', '-p', os.path.dirname(core_path)])
+    output = check_output(['sudo', 'unsquashfs', '-d', core_path, core_snap])
```
I'm curious: why unsquash versus mounting? Is that not possible from within lxc confinement?

sergiusens replied:
```
sergiusens@snapcraft:~/vim$ sudo mount /home/sergiusens/.cache/snapcraft/projects/snapcraft-core/snap_hashes/amd64/af028f2f34d9a0fbc399a36d1809e509abba1258ec9fe15d998fc87bcb779c90bc1fae7d857a3bab9eb710ddc7f4bfaa /snap/core/current
mount: /snap/core/current: mount failed: Unknown error -1
sergiusens@snapcraft:~/vim$ file /home/sergiusens/.cache/snapcraft/projects/snapcraft-core/snap_hashes/amd64/af028f2f34d9a0fbc399a36d1809e509abba1258ec9fe15d998fc87bcb779c90bc1fae7d857a3bab9eb710ddc7f4bfaa
/home/sergiusens/.cache/snapcraft/projects/snapcraft-core/snap_hashes/amd64/af028f2f34d9a0fbc399a36d1809e509abba1258ec9fe15d998fc87bcb779c90bc1fae7d857a3bab9eb710ddc7f4bfaa: Squashfs filesystem, little endian, version 4.0, 79831621 bytes, 12340 inodes, blocksize: 131072 bytes, created: Fri Jan 13 20:42:12 2017
```

sergiusens replied:
`mount` is reserved in containers and only allowed if there is a fuse interface for it

kyrofa replied:
Yeah that sounds familiar, thanks for the sanity check.

kyrofa commented on snapcraft/_store.py:
```diff
@@ -623,11 +623,23 @@ def close(snap_name, channel_names):
+    :param str download_path: the path to write the downloaded snap to.
+    :param str arch: the architecture of the download as a deb arch.
+    :param str but_hash: a sha3_384 hash to indicate that the download should
+                         not take place.
```
This variable name and description don't make sense to me. I don't understand what it does.

kyrofa commented on snapcraft/internal/lifecycle.py:
```diff
@@ -112,6 +118,42 @@ def execute(step, project_options, part_names=None):
+        download_path = os.path.join(d, 'core.snap')
+        download_hash = snapcraft.download('core', 'stable', download_path,
+                                           deb_arch, but_hash=current_hash)
+        print('download', download_hash, 'current', current_hash)
```
This should probably be removed.

kyrofa commented on snapcraft/internal/lxd.py:
```diff
@@ -59,6 +59,9 @@ def _create_container(self):
                 self._container_name])
+            check_call([
+                'lxc', 'config', 'set', self._container_name,
+                'environment.SNAPCRAFT_SETUP_CORE', '1'])
```
Much better.

kyrofa commented on snapcraft/internal/lifecycle.py:
```diff
@@ -112,6 +118,42 @@ def execute(step, project_options, part_names=None):
+    core_snap = snap_cache.get(deb_arch=deb_arch)
+
+    # Now unpack
+    logger.info('Setting up {!r} in {!r}'.format(core_snap, core_path))
```
Can we add "may prompt for sudo" here?

sergiusens replied:
I will let you log a bug so we do the same thing for every place we already call sudo ;-)

sergiusens commented:
Addressing your points:

*Make the core mount point configurable*

This was my only standing thought except it is not something I like... :-/

* Change the rmdir/mkdir/unsquashfs to not use sudo unless it gets a permissions error*

It needs to have the right permissions, so this is not a good idea.

*Create a unit test that creates a "core" snap that contains a single file, place it in the cache, and then you should be able to "setup core" and verify that it happened as expected.*

It seems too much side stepping just to run unsquashfs (and the integration test already does this).

sergiusens commented:
@elopio @elopio @elopio!!!!!

elopio commented:
It's failing everywhere! 
snapcraft.tests.test_lifecycle.CoreSetupTestCase.test_core_setup_skipped_if_core_exists
OSError: classic confinement requires the core snap to be installed. Install it by running `snap install core`.

elopio commented on snapcraft/tests/test_lifecycle.py:
```diff
@@ -607,3 +617,109 @@ def test_pull_is_dirty_if_target_arch_changes(
+        self.tempdir_mock = patcher.start()
+        self.addCleanup(patcher.stop)
+
+        # Create a fake sudo that just echos
```
this looks like a good fixture for fixture_setup.py

elopio commented:
I'm +1, once it's all green. Thanks Sergio.

sergiusens commented:
@elopio then add an approve so i don't need to ping you again, or is that what you want?

elopio commented:
This +1 is a high level of trust <3 :D
